File: shuttle-api/src/ShuttleApi.Domain/Trips/TripPreInspection.cs
using ShuttleApi.Domain.Common;

namespace ShuttleApi.Domain.Trips;

public enum FuelLevel { Full, ThreeQuarters, Half, Quarter }

public sealed class TripPreInspection : Entity<Guid>
{
    private readonly List<TripInspectionItem> _items = [];

    public Guid TripId { get; private set; }
    public int OdometerStart { get; private set; }
    public FuelLevel FuelLevel { get; private set; }
    public string? WeatherType { get; private set; }
    public string? Temperature { get; private set; }
    public string? RoadConditions { get; private set; }
    public string? Visibility { get; private set; }
    public string? RoadAdvisories { get; private set; }
    public DateTime? WeatherPulledAt { get; private set; }
    public DateTime SubmittedAt { get; private set; }
    public IReadOnlyList<TripInspectionItem> Items => _items.AsReadOnly();

    private TripPreInspection() { }

    public static TripPreInspection Create(
        Guid tripId,
        int odometerStart,
        FuelLevel fuelLevel,
        string? weatherType,
        string? temperature,
        string? roadConditions,
        string? visibility,
        string? roadAdvisories,
        DateTime? weatherPulledAt,
        IEnumerable<(string ItemName, InspectionCategory Category, bool Passed, string? Notes)> items)
    {
        var inspection = new TripPreInspection
        {
            Id = Guid.NewGuid(),
            TripId = tripId,
            OdometerStart = odometerStart,
            FuelLevel = fuelLevel,
            WeatherType = weatherType,
            Temperature = temperature,
            RoadConditions = roadConditions,
            Visibility = visibility,
            RoadAdvisories = roadAdvisories,
            WeatherPulledAt = weatherPulledAt,
            SubmittedAt = DateTime.UtcNow
        };

        foreach (var (itemName, category, passed, notes) in items)
            inspection._items.Add(TripInspectionItem.Create(inspection.Id, itemName, category, passed, notes));

        return inspection;
    }
}

File: shuttle_tablet/lib/features/trips/domain/entities/trip_pre_inspection.dart
import 'package:equatable/equatable.dart';
import 'trip_inspection_item.dart';

enum FuelLevel { full, threeQuarters, half, quarter }

class TripPreInspection extends Equatable {
  final String id;
  final String tripId;
  final int odometerStart;
  final FuelLevel fuelLevel;
  final String? weatherType;
  final String? temperature;
  final String? roadConditions;
  final String? visibility;
  final String? roadAdvisories;
  final DateTime? weatherPulledAt;
  final DateTime submittedAt;
  final List<TripInspectionItem> items;

  const TripPreInspection({
    required this.id,
    required this.tripId,
    required this.odometerStart,
    this.fuelLevel = FuelLevel.full,
    this.weatherType,
    this.temperature,
    this.roadConditions,
    this.visibility,
    this.roadAdvisories,
    this.weatherPulledAt,
    required this.submittedAt,
    this.items = const [],
  });

  @override
  List<Object?> get props => [
        id,
        tripId,
        odometerStart,
        fuelLevel,
        weatherType,
        temperature,
        roadConditions,
        visibility,
        roadAdvisories,
        weatherPulledAt,
        submittedAt,
        items,
      ];
}
